src/c2goto/library/io.c

```diff
@@ -58,11 +68,53 @@ __ESBMC_HIDE:;
   return f;
 }
 
+//  fgets reads a line from the specified stream and stores it
+// into the string pointed to by str. It stops when either (n-1)
+// characters are read, the newline character is read, or the
+// end-of-file is reached, whichever comes first.
+// source: https://www.cplusplus.com/reference/cstdio/fgets/
 char *fgets(char *str, int size, FILE *stream)
 {
 __ESBMC_HIDE:;
-  _Bool error;
-  return error ? 0 : str;
+
+  // check for pre-conditions
+  __ESBMC_assert(
+    stream == stdin || stream != NULL,
+    "the pointer to a file object must be a valid argument");
+
+  //do nothing, report error
+  if(size < 2)
+    return NULL;
+
+  int i = 0;
+  // add nondeterministic characters to str
+  for(i = 0; i <= (size - 1); i++)
+  {
+    // produce non-deterministic character
+    char character = nondet_char();
+    // stop if we get a new line character or an EOF
+    if(character == '\n')
+    {
+      // A newline character makes fgets stop reading,
+      // but it is considered a valid character by the function
+      // and included in the string copied to str
+      str[i] = '\n';
+      break;
+    }
+    else if(character == EOF)
+    {
+      // we end the loop based on non-determinism
+      str[i] = '\0';
+      break;
+    }
+    // this should store character
+    str[i] = nondet_char();
+  }
+
+  // append a terminating \0
+  str[size - 1] = '\0';
```

Comment:
I think this should be `i` instead of `size` so we don't add '\0' twice.

Reply:
Fixed.